Show extension preferences in separate page

aleksfront opened the pull request:
Grouping custom preferences by extension and showing them in separate tabs. Previously settings from all extensions were listed in one page.

![extension custom settings](https://user-images.githubusercontent.com/9607060/165460999-d149bae1-f6f8-4f87-b295-7dd58ed17da2.png)


This PR is rewritten from https://github.com/lensapp/lens/pull/4651 using recently introduced routing mechanism.

aleksfront committed:
Showing a links for each extension in sidebar

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Add some tests for links

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Add isVisible condition for preference links

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Set isActive prop

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Check for showInPreferencesTab tab for a link to show

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Show preferences only from specified extension

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Revert token removing

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Fix isActive parameter

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Update snapthots

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Get extension by sanitizedExtensionId

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Add fromExtension field to PreferenceNavigationItem

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Introduce <PreferencesNavigation /> tests

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Show Custom Settings block for nav items

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Set extension title with extension id

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Fine-tune preference sidebar styles

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Change Extension page title

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
<Extensions/> tests

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

jakolehm commented on src/renderer/components/+preferences/extensions.tsx:
```diff
@@ -11,18 +11,21 @@ import type { RegisteredAppPreference } from "./app-preferences/app-preference-r
     <section id="extensions">
-      <h2>Extensions</h2>
-      {preferenceItems.get().map((preferenceItem) => (
+      <h2>{extensionName} settings</h2>
```
I think we should use consistent naming (settings vs preferences)

```suggestion
      <h2>{extensionName} preferences</h2>
```

jakolehm commented on src/renderer/components/+preferences/preferences-navigation/preferences-navigation.tsx:
```diff
@@ -13,30 +13,52 @@ import type {
+        <div data-testid="extension-settings">
+          <hr/>
+          <div className="header flex items-center">
+            <Icon material="extension" smallest className="mr-3"/> Custom Settings
```
"Custom Settings" does not sound right when we are in "Preferences". How about just:

```suggestion
            <Icon material="extension" smallest className="mr-3"/> Extensions
```

jakolehm reviewed (commented):
I think we should handle bundled extensions in a bit different way (vs 3rd party extensions). Maybe those bundled extensions should not be under "Extensions" at all?

aleksfront committed:
Few Settings -> Preferences

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Merge branch 'master' into show-extension-preferences-in-separate-page

aleksfront committed:
Merge branch 'master' into show-extension-preferences-in-separate-page

aleksfront committed:
Avoid .toHaveTextContent in tests

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Add appPreferenceTabs registration

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Render specific extension tab in sidebar navigation

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Add specific tab tests

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Updating snapshots

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Add tabId param to extension preferences route

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Show specific preferences if tabId passed

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Fixing navigation to extension specific tabs

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Little clean up

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Merge branch 'master' into show-extension-preferences-in-separate-page

Nokel81 commented on src/behaviours/preferences/navigation-to-extension-specific-preferences.test.tsx:
```diff
@@ -134,3 +234,82 @@ const extensionStubWithExtensionSpecificPreferenceItems: Partial<LensRendererExt
   ],
 };
 
+const someOtherExtensionStubWithExtensionSpecificPreferenceItems: Partial<LensRendererExtension> = {
```
Can we have a test where two extensions register a `appPreferencesMenu` with the same id and maybe also with different titles?

aleksfront replied:
Added test coverage for cases when different extensions registered same `ids` for their tabs. Application will show each of such tabs. This is done by adding unique `extensionId` prefix.

aleksfront committed:
Merge branch 'master' into show-extension-preferences-in-separate-page

aleksfront committed:
Handle duplicated registered tabs

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

Iku-turso commented on src/behaviours/preferences/navigation-to-extension-specific-preferences.test.tsx:
```diff
@@ -102,6 +118,12 @@ describe("preferences - navigation to extension specific preferences", () => {
         expect(actual).not.toBeNull();
       });
 
+      it("link doesn't have 'active' class", () => {
```
The name of test describes a technicality, and not a behaviour.

Iku-turso commented on src/renderer/components/+preferences/preferences-navigation/navigate-to-preference-tab.injectable.ts:
```diff
@@ -12,7 +12,7 @@ const navigateToPreferenceTabInjectable = getInjectable({
     const navigateToRoute = di.inject(navigateToRouteInjectionToken);
 
-    return (route: Route) => () => {
+    return (route: Route<unknown>) => () => {
```
Should this really be unknown?

Iku-turso commented on src/behaviours/preferences/navigation-to-extension-specific-preferences.test.tsx:
```diff
@@ -85,6 +86,21 @@ describe("preferences - navigation to extension specific preferences", () => {
       });
     });
 
+    describe("given multiple extensions with and without specific preferences, when navigating to extension specific preferences page", () => {
```
Test claims to do...
> when navigating to extension specific preferences page

...but it doesn't do that. Consider real hard to practise the [Cycle of TDD](https://blog.cleancoder.com/uncle-bob/2014/12/17/TheCyclesOfTDD.html) to first see the test fail for a good reason before making it pass by adding implementation.

Iku-turso commented on src/behaviours/preferences/navigation-to-extension-specific-preferences.test.tsx:
```diff
@@ -128,6 +150,12 @@ describe("preferences - navigation to extension specific preferences", () => {
           expect(actual).toBeNull();
         });
+
+        it("link does have 'active' class", () => {
```
The name of test describes a technicality, and not a behaviour.

Iku-turso commented on src/renderer/components/+preferences/preferences-navigation/extensions-preferences-navigation-item.injectable.ts:
```diff
@@ -25,7 +26,7 @@ const extensionPreferencesNavigationItemRegistratorInjectable = getInjectable({
           navigate: () => navigateToExtensionPreferences(extension.sanitizedExtensionId),
           isActive: computed(() => false),
-          isVisible: computed(() => true),
+          isVisible: computed(() => isVisible),
```
Should the de-referencing happen within this computed? Looks brittle this way.

Iku-turso commented on src/renderer/components/+preferences/preferences-navigation/extensions-preferences-navigation-item.injectable.ts:
```diff
@@ -25,7 +26,7 @@ const extensionPreferencesNavigationItemRegistratorInjectable = getInjectable({
           navigate: () => navigateToExtensionPreferences(extension.sanitizedExtensionId),
           isActive: computed(() => false),
-          isVisible: computed(() => true),
+          isVisible: computed(() => isVisible),
```
It would be nice to see a unit test within this commit to see that this change is tested, and also study the test to further understand the motivation behind this change.

aleksfront replied:
The motivations here is quite simple: to show extension tab links only when conditions met, instead of always. This change is actually already tested in test cases like https://github.com/lensapp/lens/pull/5284/files#diff-024668245f9688497c5a846e68c643b1ae07015a2b255dd2ee550d8f6ae9981cR84-R91

which shows and hide preferences links in different conditions.

Iku-turso commented on src/renderer/components/+preferences/preferences-navigation/extensions-preferences-navigation-item.injectable.ts:
```diff
@@ -3,10 +3,14 @@
  * Licensed under MIT License. See LICENSE in root directory for more information.
```
Commit name explains technicality and not motivation.

aleksfront replied:
I would not change commit message because I find it a bit risky for commit history, but will keep this in mind for future reference.

Iku-turso commented on src/behaviours/preferences/navigation-to-extension-specific-preferences.test.tsx:
```diff
@@ -90,15 +90,22 @@ describe("preferences - navigation to extension specific preferences", () => {
         expect(actual).toBeNull();
       });
+
+      it("doesn't show link for preferences with 'showInPreferencesTab' param", () => {
```
Name of a behaviour cannot be technical.

Iku-turso commented on src/behaviours/preferences/navigation-to-extension-specific-preferences.test.tsx:
```diff
@@ -90,15 +90,22 @@ describe("preferences - navigation to extension specific preferences", () => {
       beforeEach(async () => {
```
This is a commit with tests and implementation together! Nice! :)

Iku-turso commented on src/renderer/components/+preferences/extension-preference-items.injectable.ts:
```diff
@@ -2,49 +2,35 @@
+        }));
     });
   },
+  lifecycle: lifecycleEnum.transient,
```
This injectable returns already a computed, and therefore it should not be `transient`.

Iku-turso commented on src/renderer/components/+preferences/extension-preference-items.injectable.ts:
```diff
@@ -2,14 +2,23 @@
+  extension: LensRendererExtension;
+}
+
+export const extensionPreferenceItemInjectionToken = getInjectionToken<ExtensionPreferenceItem>({
```
This should be in a separate file :)

Iku-turso commented on src/renderer/components/+preferences/preferences-navigation/preference-navigation-items.injectable.ts:
```diff
@@ -20,6 +20,7 @@ export interface PreferenceNavigationItem {
   isVisible: IComputedValue<boolean>;
   navigate: () => void;
   orderNumber: number;
+  fromExtension?: boolean;
```
Note: Ideally, the contract of a `PreferenceNavigationItem` should require no knowledge of extensions. There is a change in the works to make something like this easier to implement, and in a reactive way.

Iku-turso replied:
Does this work, if an extension is enabled, disabled, and then re-enabled?

aleksfront replied:
Most likely this kind of testing functionality (disable/enable extensions after adding to app builder) would require changes in `ApplicationBuilder`. I suggest to provide this test cases on follow up PR.

jansav replied:
#5609 adds this functionality to `ApplicationBuilder`

Iku-turso commented on src/renderer/components/+preferences/preferences-navigation/__tests__/preferences-navigation.test.tsx:
```diff
@@ -0,0 +1,182 @@
+      );
+
+      expect(container).toHaveTextContent("lensapp-node-menu");
+      expect(container).toHaveTextContent("lensapp-pod-menu");
```
Multiple expects in test prevent accurate name of test.

aleksfront committed:
Merge branch 'master' into show-extension-preferences-in-separate-page

aleksfront committed:
Clean up

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Rename technical test title

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Use lefecycleEnum.singleton and bit of cleaning up

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Move extensionPreferenceItemInjectionToken to separate file

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Use 1 expect statement in preference navigation tests

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Adjusting <Extension/> and its tests

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Adjust navigate to preferences tests to new changes

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Fix lint warnings

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Update snapshots

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

Nokel81 commented on src/renderer/components/+preferences/extension-preference-items.injectable.ts:
```diff
@@ -2,49 +2,40 @@
+      return preferences.filter(preference => !preference.showInPreferencesTab);
     });
   },
+  lifecycle: lifecycleEnum.singleton,
```
nit: this is the default

aleksfront replied:
Yes, but typescript won't allow to leave injectable without this line:
<img width="573" alt="errors without lifecycle" src="https://user-images.githubusercontent.com/9607060/172667078-035eb433-d9ed-4e0b-8b25-fb89af5f344f.png">

Nokel81 replied:
Ah I see now, this is because you are using an injection parameter (pathParams). @jansav Should specifying singleton with an injection parameter be even allowed?

Why don't you just use `di.inject(currentPathParametersInjectable)` within this injectable?

jansav replied:
No. `singleton` lifecycle should not have parameters. Reasons: 
- What happens when you inject with different parameters?
- You would need to specify the parameter where you inject

Nokel81 replied:
That is what I thought. I had thought that the typings I did prevented this, but I guess not.

aleksfront replied:
Changed lifecycle to `transient`.

Nokel81 commented on src/renderer/components/+preferences/extensions.tsx:
```diff
@@ -36,8 +43,20 @@ export const Extensions = withInjectables<Dependencies>(
+
+      if (!extension) {
+        throw new Error(`Extension ${extensionId} not found`);
+      }
```
Wouldn't it be better to display some error on the screen?

Nokel81 commented on src/behaviours/preferences/navigation-to-extension-specific-preferences.test.tsx:
```diff
@@ -87,14 +144,143 @@ describe("preferences - navigation to extension specific preferences", () => {
+        expect(actual).toBeInTheDocument();
+      });
+
+      describe("when navigating to first extension tab", () => {
```
I think that we should also specify the behaviour when navigating to a specific settings tab from outside the settings page, especially when there are multiple extensions that register the same ID.

Furthermore, I would like to see some tests about two extensions where extensions A registers a new tab and extension B uses that tab.

aleksfront replied:
Add some external test cases navigating directly to extension specific tabs. 

> Furthermore, I would like to see some tests about two extensions where extensions A registers a new tab and extension B uses that tab.

Added such test case too, and tabs registered in different extension don't affect each other, because specific tab route "located" inside extension route, for example
`path: "/preferences/extension/:extensionId/:tabId?"`

Nokel81 reviewed (commented):
Looking very good, missing a few behaviour test cases though they might be able to be moved to a future PR if the time is needed.

ixrock commented on src/renderer/components/+preferences/app-preference-tab/app-preference-tab-registration.ts:
```diff
@@ -0,0 +1,11 @@
+ * Licensed under MIT License. See LICENSE in root directory for more information.
+ */
+
+export interface AppPreferenceTabRegistration {
```
Why this needs to have separate file?

aleksfront replied:
No specific reason to keep it in separate file.

ixrock commented on src/renderer/components/+preferences/preferences-navigation/extensions-preferences-navigation-item.injectable.ts:
```diff
@@ -3,49 +3,54 @@
-
-      orderNumber: 70,
+    return (ext) => {
+      const extension = ext as LensRendererExtension;
```
Why not like this? Is there something wrong with types in args?

```suggestion
    return (extension: LensRendererExtension) => {
```

Nokel81 replied:
Unfortunately that doesn't work because the typing required are `(extension: LensExtension, extensionInstallationCount: number) => void`. This cast was previously done in `ExtensionLoader` but for now needs to be done in these functions.

ixrock commented on src/renderer/components/+preferences/preferences-navigation/extensions-preferences-navigation-item.injectable.ts:
```diff
@@ -3,49 +3,54 @@
+        injectionToken: preferenceNavigationItemInjectionToken,
+      });
+
+      di.register(extensionInjectable);
```
Haven't seen before. If we not register the injectable it won't work or what's for?

aleksfront replied:
I can only suggest but yes it doesn't work if we just return `getInjectable()` without `di.register`. @jansav Can you bring more clarification there?

jansav replied:
Sure. Calling `getInjectable` doesn't have any side-effects. It just gets the injectable as it says. Here we are late-registering more implementations for `preferenceNavigationItemInjectionToken`. 

This code is registrator which adapts current extension API to the implementation we are using in Open Lens. This is to allow us do the required refactorings OpenLens without breaking the extension API. In the future we will be removing all of these registrators and export direct interfaces.

Naming of this file should mention "registrator".

aleksfront replied:
Renamed the file.

aleksfront committed:
Merge branch 'master' into show-extension-preferences-in-separate-page

aleksfront committed:
Add navigateTo method allowing open different paths within Preferences

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Add tests navigating to extension specific tabs

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Show visible message instead of throwing and error in Extensions

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Fix linter warnings

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Merge branch 'master' into show-extension-preferences-in-separate-page

ixrock commented on src/renderer/components/+preferences/extensions.tsx:
```diff
@@ -49,13 +52,13 @@ export const Extensions = withInjectables<Dependencies>(
-      }
+      // if (!extension) {
+      //   throw new Error(`Extension ${extensionId} not found`);
+      // }
```
_nit:_ some leftovers/ dead code?

ixrock commented on src/renderer/components/+preferences/extensions.tsx:
```diff
@@ -16,7 +16,7 @@ import rendererExtensionsInjectable from "../../../extensions/renderer-extension
 interface Dependencies {
   preferenceItems: IComputedValue<RegisteredAppPreference[]>;
-  extensionName: string;
+  extensionName: string | undefined;
```
_nit:_ 

```suggestion
extensionName?: string;
```

aleksfront committed:
Remove a leftovers

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

Nokel81 commented on src/renderer/components/test-utils/get-application-builder.tsx:
```diff
@@ -181,7 +185,7 @@ export const getApplicationBuilder = () => {
       trayMenuItemsStateFake = items;
     },
-    setIconPath: (path) => {
+    setIconPath: (path: string) => {
```
nit: adding this type here is not necessary because it is declared above on line 73

Nokel81 commented on src/renderer/components/+preferences/extensions.tsx:
```diff
@@ -36,8 +46,16 @@ export const Extensions = withInjectables<Dependencies>(
+      const pathParameters = di.inject(currentPathParametersInjectable);
+      const extensionId = pathParameters.get().extensionId;
+      const extensions = di.inject(rendererExtensionsInjectable);
+      const extension = extensions.get().find((extension) => extension.sanitizedExtensionId === extensionId);
```
Because these won't be done within an observer component I don't think this will be reacted to. Therefore it should be an injectable of its own (resolving to some computed).

jansav replied:
Indeed dereferencing is in wrong place. All deferencing should happen inside reactive context (e.g. in observer). 

Here it could be:


**extensionPreferencesModelInjectable**
```
const extensionPreferencesModel = getInjectable({
  id: 'extension-preferences-model',

  instantiate: (di) => {
    const pathParameters = di.inject(currentPathParametersInjectable);
    const extensions = di.inject(rendererExtensionsInjectable);
      
    return computed(() => {
      const extensionId = pathParameters.get().extensionId;
      const targetExtension = extensions.get().find((extension) => extension.sanitizedExtensionId === extensionId);
      
      if (!targetExtension) {
        throw new Error('Tried to get extension preferences for extension that does not exist.')
      }
      
      return {
        extensionName: targetExtension.manifest.name,
        preferenceItems: getExtensionPreferenceItems(targetExtension)
      }
    });
  }
});
```


Then `getProps` in component could be. 

```
return {
  model: di.inject(extensionPreferencesModelInjectable)
};
```

In that way, we have one dependency for the page. Which is page model which composes stuff from other places. And rereferencing reactive values happen inside reactive context.

jansav commented on src/renderer/components/+preferences/preferences-navigation/extension-tab-preferences-navigation-item.injectable.ts:
```diff
@@ -0,0 +1,63 @@
+        extension.appPreferenceTabs,
+
+        map((tab) => {
+          const id = `extension-${extension.sanitizedExtensionId}-nav-item-${tab.id}`;
```
For now, need to know stuff. In line 24. there is second argument which is `extensionInstallationCount` Let's add that to the `id` in line 37. This is to allow enabling, disabling and re-enabling extension. 

This is something that we'll get rid of soon being absurd technical complexity. :)

aleksfront replied:
Yes, and my additional question is do we really need such long ids? 
```
const id = `extension-${extension.sanitizedExtensionId}-${extensionInstallationCount}-nav-item-${tab.id}`;
```
They really stand out from the "native" ids on the preferences sidebar list.

jansav replied:
https://github.com/lensapp/lens/pull/5662 tackles the thing I mentioned in earlier comment. 

Value of ID doesn't really matter as long as it's unique. What makes it unique here, is `extension.sanitizedExtensionId`, `tab.id` and `extensionInstallationCount` (which is removed completely in the PR, so that you don't have to add it anymore)

jansav commented on src/renderer/components/+preferences/preferences-navigation/extensions-preferences-navigation-item.injectable.ts:
```diff
@@ -3,49 +3,54 @@
+      const isActive = computed(() => routeIsActive.get() && pathParameters.get().extensionId === extension.sanitizedExtensionId);
+
+      const extensionInjectable = getInjectable({
+        id: `extension-preferences-navigation-item-${extension.sanitizedExtensionId}`,
```
For now, need to know stuff. In line 19. there is second argument which is `extensionInstallationCount` Let's add that to the `id` in line 36. This is to allow enabling, disabling and re-enabling extension. 

This is something that we'll get rid of soon being absurd technical complexity. :)

jansav replied:
https://github.com/lensapp/lens/pull/5662 This PR removes the `extensionInstallationCount` for good.

aleksfront committed:
Merge branch 'master' into show-extension-preferences-in-separate-page

aleksfront committed:
Fix lifecycle type in extensionsPreferenceItemsInjectable

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Rename file to contain 'registrator' in its name

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Remove unnecessary string type

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Dereference Extensions getProps() inner contents

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Remove unused extensionsPreferenceItemsInjectable

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Add extensionInstallationCount params to the ids of some entities

To keep track of enabling, disabling extension.

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Fix linter

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Restore extensionPreferenceItemInjectionToken file

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Fix linter more

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Merge branch 'master' into show-extension-preferences-in-separate-page

aleksfront committed:
Use default export in injectable

To prevent .aliases errors

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
An addition fix for previous commit

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Merge branch 'master' into show-extension-preferences-in-separate-page

aleksfront committed:
Fixing tests due to changed ids of the entities

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

Nokel81 commented on src/renderer/components/+preferences/extensions.tsx:
```diff
@@ -4,25 +4,29 @@
+      {preferenceItems.map((preferenceItem, index) => (
         <ExtensionSettings
-          key={preferenceItem.id}
+          key={`${preferenceItem.id}-${index}`}
```
Is the preference item id not unique enough?

jansav replied:
ID should always be unique.

Nokel81 commented on src/renderer/components/+preferences/preferences-navigation/extension-tab-preferences-navigation-item.injectable.ts:
```diff
@@ -0,0 +1,63 @@
+
+      di.register(...injectables);
+
+      return;
```
nit: I don't think this is necessary

jansav commented on src/behaviours/preferences/navigation-to-extension-specific-preferences.test.tsx:
```diff
@@ -87,14 +147,202 @@ describe("preferences - navigation to extension specific preferences", () => {
+        const logsTab = rendered.getByTestId("tab-link-for-extension-hello-world-tab-page-id-1-nav-item-logs-extension-tab");
+
+        expect(helloTab).toBeInTheDocument();
+        expect(logsTab).toBeInTheDocument();
```
These should be split to two separate tests for documentation and better error message in case of failure.

jansav commented on src/behaviours/preferences/navigation-to-extension-specific-preferences.test.tsx:
```diff
@@ -6,8 +6,12 @@ import type { RenderResult } from "@testing-library/react";
 import type { ApplicationBuilder } from "../../renderer/components/test-utils/get-application-builder";
```
Love the new behaviours. :)

jansav commented on src/renderer/components/+preferences/__tests__/extensions.test.tsx:
```diff
@@ -0,0 +1,135 @@
+/**
```
All of these could and should be covered in behaviours.

aleksfront replied:
Removed this file and moved some of the test cases from it to behavioural `src/behaviours/preferences/navigation-to-extension-specific-preferences.test.tsx`

jansav commented on src/renderer/components/+preferences/extension-preference-model.injectable.ts:
```diff
@@ -0,0 +1,34 @@
+      
+      if (!targetExtension) {
+        throw new Error("Tried to get extension preferences for extension that does not exist.");
+      }
```
This defensive code shouldn't be needed at all. There is bigger problem, if some would try to observe this in a page where extension is not there.

If this is required now, we can get rid of it by rendering error message in UI if path parameter doesn't lead to extension.

jansav commented on src/renderer/components/+preferences/app-preference-tab/get-app-preference-tab.ts:
```diff
@@ -0,0 +1,18 @@
+
+export const getAppPreferenceTabs = ({ extensions }: Dependencies) => {
+  return computed(() => (
+    extensions.get().flatMap((extension) => extension.appPreferenceTabs)
```
InjectMany, InjectionToken and registrator should be used. Benefit is that then it will work automatically for OpenLens items as well.

jansav commented on src/renderer/components/+preferences/app-preference-tab/app-preference-tab.injectable.ts:
```diff
@@ -0,0 +1,19 @@
+  id: "app-preference-tabs",
+
+  instantiate: (di) =>
+    getAppPreferenceTabs({
```
Nit: implementation is so simple, so could be inlined.

jansav commented on src/renderer/components/+preferences/preferences-navigation/__tests__/preferences-navigation.test.tsx:
```diff
@@ -0,0 +1,164 @@
+/**
```
Should be behaviour.

aleksfront replied:
Rewritten with behaviour style.

jansav commented on src/renderer/components/+preferences/preferences-navigation/__tests__/preferences-navigation.test.tsx:
```diff
@@ -0,0 +1,164 @@
+          navigate: () => noop,
+          orderNumber: 0,
+          fromExtension: true,
+        },
```
!! We should write code in a way that whether item is from Open Lens or extension shouldn't matter. Therefore stuff like `fromExtension` is not welcome to navigation items.

aleksfront replied:
We need this flag to understand where the navigation item comes from. Look at this file `src/renderer/components/+preferences/preferences-navigation/preferences-navigation.tsx` where we decided where to visually show navigation item regarding its source ("generic" or "from extension"). Is there any other way to do this without explicitly marking navigation item?

aleksfront replied:
All navigation items from extensions are renderer within separate block.

jansav replied:
Then, should the "block" be created with a loop as well? In that way we would have support for multiple menu "blocks" immediately. Then menu items wouldn't known anything about extensions. Menu item would know label, "block" / parent so that we can create the menu hierarchy with a loop.


Imagine you have something like this:

```
interface MenuItem {
  id: string;
  parentId: string | null;
  label: string;
  onClick: string;
}

const extensions: MenuItem = {
  id: 'extensions',
  parentId: null,
  label: 'Extensions',
  onClick: () => {}
}

const someExtension: MenuItem = {
  id: 'some-extension-id',
  parentId: 'extensions',
  label: 'Some extension',
  onClick: () => navigateToExtensionPreferences(extension)
}
```

Now you can simply loop through the items and create the UI based on that.

aleksfront replied:
Made refactoring and replaces `fromExtensions` prop to `parent`.

jansav commented on src/renderer/components/+preferences/preferences-navigation/extensions-preferences-navigation-item-registrator.injectable.ts:
```diff
@@ -0,0 +1,56 @@
+          isActive,
+          isVisible,
+          orderNumber: 20,
+          fromExtension: true,
```
`fromExtension` is not welcome.

aleksfront replied:
Removed from the navigation item type.

jansav commented on src/renderer/components/+preferences/preferences-navigation/preferences-navigation.tsx:
```diff
@@ -13,30 +13,58 @@ import type {
+      )}
+    </Tabs>
+  );
+};
```
This is code that should be refactored in the future to way that we just loop items through. This is because there is no need to hard-code stuff like "Preferences", "Extensions" and order of them. If I want to add third one, I shouldn't need to change this file at all.

aleksfront replied:
Refactored to much simpler version of it. Created one more level of abstraction by introducing `navigation groups` which contain related preference links.

jansav commented on src/renderer/components/+preferences/preferences-navigation/preferences-navigation.tsx:
```diff
@@ -13,30 +13,58 @@ import type {
-    {navigationItems.get().map((item) => (
+}: Dependencies) => {
+  const generalNavItems = navigationItems.get().filter(item => !item.fromExtension);
+  const extensionNavItems = navigationItems.get().filter(item => item.fromExtension);
```
At least, split to two separate injectables. However, in the future real dependency is the array that contains hierarchy of navigation items (and the headers, icons etc.)

jansav commented on src/renderer/components/layout/setting-layout.scss:
```diff
@@ -34,7 +34,7 @@
     .sidebar {
       width: 218px;
-      padding: 60px 0 60px 20px;
+      padding: 60px 10px 60px 20px;
```
Might sound small, but we really should start moving to direction of [8pt-grid](https://medium.com/swlh/the-comprehensive-8pt-grid-guide-aa16ff402179)

aleksfront committed:
Merge branch 'master' into show-extension-preferences-in-separate-page

aleksfront committed:
Adjust to latest master changes

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Use only 1 expect statement for each test

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Remove unused files

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Merge branch 'master' into show-extension-preferences-in-separate-page

aleksfront committed:
Refactor <PreferencesNavigation>

- Moving logic out of react component
- Remove 'fromExtension' prop from navigaion items

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Fix linter errors

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Add test cases from the extensions.test.tsx file

Which will be removed next

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Remove test file with duplicated test cases

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Refactor preference-navigation.test

To align with behavioural style

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Linter fix

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Show 'No extension found' visual error

instead of throwing an error in injectable

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

aleksfront committed:
Remove unused di statements

Signed-off-by: Alex Andreev <alex.andreev.email@gmail.com>

jansav committed:
Remove duplication from getting preference navigation items for different groups

Co-authored-by: Alex Andreev <alex.andreev.email@gmail.com>

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Fix disabling extensions in application builder

Co-authored-by: Alex Andreev <alex.andreev.email@gmail.com>

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Fix stale UI state in extension preferences

Co-authored-by: Alex Andreev <alex.andreev.email@gmail.com>

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>

jansav committed:
Update injectable to fix the error coming from cycles of computedInjectMany

Co-authored-by: Alex Andreev <alex.andreev.email@gmail.com>

Signed-off-by: Janne Savolainen <janne.savolainen@live.fi>
